File: 1095_findInMountainArray.py
# """
# This is MountainArray's API interface.
# You should not implement it, or speculate about its implementation
# """
#class MountainArray:
#    def get(self, index: int) -> int:
#    def length(self) -> int:

class Solution:
    def findInMountainArray(self, target: int, mountain_arr: 'MountainArray') -> int:
        index = self.findPeakInMountainArray(mountain_arr)
        if mountain_arr.get(index) == target:
            return index
        idx = self.binarySearch(target, mountain_arr, 0, index - 1, False)
        return idx if idx != -1 else self.binarySearch(target, mountain_arr, index + 1, mountain_arr.length() - 1, True)
        
    def findPeakInMountainArray(self, mountain_arr):
        length = mountain_arr.length()
        left, right = 0, length - 1
        while left < right:
            mid = left + (right - left) // 2
            if mountain_arr.get(mid + 1) > mountain_arr.get(mid):
                left = mid + 1
            else:
                right = mid
        return left
        
    def binarySearch(self, target, mountain_arr, left, right, reverse):
        while left <= right:
            mid = left + (right - left) // 2
            val = mountain_arr.get(mid)
            if val == target:
                return mid
            elif (not reverse and val > target) or (reverse and val < target):
                right = mid - 1
            else:
                left = mid + 1
        return -1

File: 1001-1500/1095_findInMountainArray.java
/**
 * // This is MountainArray's API interface.
 * // You should not implement it, or speculate about its implementation
 * interface MountainArray {
 *     public int get(int index) {}
 *     public int length() {}
 * }
 */
 
class Solution {
    public int findInMountainArray(int target, MountainArray mountainArr) {
        int length = mountainArr.length();
        int low = 0; int high = length - 1; int peek = -1;
        int mid; int prev; int curr; int next;

        while (low <= high) {
            mid = low + (high - low) / 2;

            prev = mountainArr.get(mid - 1);
            curr = mountainArr.get(mid);
            next = mountainArr.get(mid + 1);

            if (curr > prev && curr > next) {
                peek = mid;
                break;
            } else if (prev < curr && curr < next) {
                low = mid;
            } else {
                high = mid;
            }
        }

        int result = -1;

        result = binarySearch(mountainArr, target, 0, peek, true);

        if (result != -1) {
            return result;
        }

        result = binarySearch(mountainArr, target, peek, length - 1, false);

        return result;
    }

    public int binarySearch(MountainArray mountainArr, int target, int low, int high, boolean increase) {
        int val; int mid;

        while (low <= high) {
            mid = low + (high - low) / 2;
            val = mountainArr.get(mid);

            if (val == target) {
                return mid;
            } else if ((val < target && increase) || (val > target && !increase)) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }

        return -1;
    }
}
